text/0000-safer-transmute.md

```diff
@@ -0,0 +1,1796 @@
+ - and if the destination type is a mutate-able reference, [preserve validity][reference-validity].
+
+##### Preserve or Shrink Size
+[reference-size]: #Preserve-or-Shrink-Size
+
+You may preserve or decrease the size of the referent type via transmutation:
+```rust
+let _: &[u8; 3] = (&[0u8; 9]).transmute_into();
+```
+
+However, you may **not**, under any circumstances, *increase* the size of the referent type:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+let _: &[u8; 9] = (&[0u8; 3]).transmute_into(); // Compile Error! 
+```
+##### Preserve or Relax Alignment
+[reference-alignment]: #Preserve-or-Relax-Alignment
+
+Unaligned loads are undefined behavior. You may transmute a reference into reference of more relaxed alignment:
+```rust
+let _: &[u8; 0] = (&[0u16; 0]).transmute_into();
+```
+
+However, you may **not** transmute a reference into a reference of more-restrictive alignment:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+let _: &[u16; 0] = (&[0u8; 0]).transmute_into(); // Compile Error! 
+```
+
+##### Preserve or Shrink Lifetimes
+[reference-lifetimes]: #Preserve-or-Shrink-Lifetimes
+
+You may transmute a reference into a reference of lesser lifetime:
+```rust
+fn shrink<'a>() -> &'a u8 {
+    static long : &'static u8 = &16;
+    long.transmute_into()
+}
+```
+
+However, you may **not** transmute a reference into a reference of greater lifetime:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+fn extend<'a>(short: &'a u8) -> &'static u8 {
+    short.transmute_into() // Compile Error!
+}
+```
+
+##### Preserve or Shrink Uniqueness
+[reference-mutability]: #Preserve-or-Shrink-Uniqueness
+
+You may preserve or decrease the uniqueness of a reference through transmutation:
+```rust
+let _: &u8 = (&42u8).transmute_into();
+let _: &u8 = (&mut 42u8).transmute_into();
+```
+
+However, you may **not** transmute a shared reference into a unique reference:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+let _: &mut u8 = (&42u8).transmute_into(); // Compile Error!
+```
+
+##### Mutate-able References Must Preserve Validity
+[reference-validity]: #Mutate-able-References-Must-Preserve-Validity
+
+A mutate-able reference is:
+- all unique (i.e., `&mut T`) references
+- all shared (i.e., `&T`) references whose referent type contain any bytes produced by the contents of `UnsafeCell`.
+
+Unlike transmutations of owned values, the transmutation of a mutate-able reference may also not expand the bit-validity of the referenced type. For instance:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+let mut x = NonZeroU8::new(42).unwrap();
+{
+    let y : &mut u8 = (&mut x).transmute_into(); // Compile Error!
+    *y = 0;
+}
+
+let z : NonZeroU8 = x;
+```
+If this example did not produce a compile error, the value of `z` would not be a bit-valid instance of its type, [`NonZeroU8`].
+
+
+
+### 📖 When is a transmutation safe?
```

Comment:
This primarily talks about structs and enums, but what about primitive types? 

Primitive types, specifically references, have some safety concerns that I don't see addressed in here anywhere (unless I'm missing something). For example:

- transmuting a `NonZeroUsize` to a `&u8` is sound (both have well-defined layout, bit-validity is preserved, and alignment is preserved), but not safe (the value could point to invalid memory).
- transmuting a `&[u8]` to a `&str` is currently unsound due to the absence of a well-defined layout, but they did have well-defined layouts then the transmutation would be sound, but not safe, since the `&[u8]` may have invalid byte sequences for a utf-8 encoding.

The easiest thing to do is to say that references can't be safely transmuted, but that is more restrictive than necessary.  A more flexible approach would be to do something like only allow transmuting types containing references, if that segment of memory in the type is a reference in both the source and destination types, and the types pointed to are transmute compatible. 

Meaning that this is fine:
```rust
#[repr(C)]
struct A(&u32);
struct B(&[u8; 4]);

let b: B = A(1u32).transmute_into();
```

But this isn't: 
```rust
#[repr(C)]
struct A(&[u8; 4]);
struct B(NonZeroUsize);

let a: A = unsafe {B(NonZeroUSize::new_unchecked(1)) }.transmute_into();
```

Reply:
> Transmuting `NonZeroUsize` to `&u8` is sound but not safe

That's an open question, actually. It's not decided whether `&T` has a validity requirement to be dereferencable yet.

IIRC, the UWG's consensus currently seems to be that `&T`'s validity requirement should be just "aligned and nonnull" and the safety requirement be "dereferences to a `T` that holds its safety requirements," but that the safety requirement of the reference is aggressively asserted whenever the reference undergoes a typed copy.

Or perhaps what the borrow model (Stacked Borrows) asserts on a copy of the reference is between those two? Just that the reference dereferences to the correct amount of memory with the correct provenance? Just that the pointee is valid, but not necessarily safe?

In any case, the exact validity/safety requirement split for references is not decided yet, so any transmutation provided by this document should keep that in mind.

Reply:
Whenever I was unsure of whether a requirement was a soundness constraint or just a safety constraint, I erred on side of soundness. (Except in my introduction of `NeglectAlignment`, which I've now made note of, and will correct quite soon.) A soundness constraint can be later relaxed into a safety constraint, but not the other way around.

Ultimately, I try to avoid getting *too* into the nitty-gritty of the transmutability rules. The primary goal of this RFC is to specify an API that's general enough to encompass whatever rules there happen to be, *not* to specify Rust's soundness and safety rules. The examples in this section are intended as illustrative descriptions, not prescriptions.

Types with exotic representations, like `NonZero*` will require some special implementation attention.